design/pipeline-resources.md

```diff
@@ -0,0 +1,190 @@
+
+```yaml
+resources:        # types: pipelines | repositories | containers | packages
+  pipelines:
+  - pipeline: string  # identifier for the pipeline resource
+    type: enum  # type of the pipeline source like azurePipelines, Jenkins etc. 
+    connection: string  # service connection to connect to the source
+    source:  
+      name: string  # source defintion of the pipeline including the project i.e. projectName/definition
+      version: string  # version to pick the artifact, optional; defaults to Latest
+      branch: string  # branch to pick the artiafct, optional; defaults to master branch
+      tags: string # picks the artifacts on from the pipeline with given tag, optional; defaults to no tags.
+```
+
+The inputs inside `source` can change based on the pipeline type defined. The above schema is for the `type`: azurePipelines.
+
+
+### Examples
+
+If you need to consume another azurePipelines from the current project and you dont require setting branch, version and tags info, this can be shortened to:
+
+```yaml
+resources:         
+  pipelines:
+  - pipeline: SmartHotel      
+    type: azurePipelines
+    source: SmartHotel-CI  # name of the pipeline source definition
+```
+
+In case you need to consume an azurePipeline from other project, then you need to include the project name while providing source name.
+
+```yaml
+resources:         
+  pipelines:
+  - pipeline: SmartHotel      
+    type: azurePipelines
+    source: 
+      name: DevOpsProject/SmartHotel-CI  # name of the pipeline source from different project
+      branch: releases/M142
+```
+
+### `downloadArtifact` for pipelines
+
+Artifacts from the `pipeline` resource are automatically downloaded and made available for all the jobs. However, in any of the jobs, you can choose to override and download only specific artifacts using `downloadArtifact` shortcut.
+
+
+```yaml
+- job: deploy_windows_x86_agent
+  steps:
+  - downloadArtifact: SmartHotel   # pipeline resource identifier.
+    name: WebTier1  # artifact to download, optional; defaults to all the artifacts from the resource.
+    patterns: '**/*.zip'  # mini match pattern to download specific files, optional; defaults to all files.
+```
+
+Or to avoid downloading any of the artifacts at all:
+
+```yaml
+- downloadArtifact: none
+```
+
+
+Refer to [download artifacts](https://github.com/Microsoft/azure-pipelines-yaml/blob/master/design/pipeline-artifacts.md#downloading-artifacts-downloadartifact) for more details.
+
+Artifacts from the `pipeline` resource are downloaded to `$PIPELINES_RESOURCESDIR/<pipeline-identifier>/` folder.
+
+## Resources: `repositories`
+
+If you have multiple repositories from which you need to sync the code into your pipeline, you can consume the repos by defining a `repositories` resource. A repository can be another Azure Repo or any external repo like GitHub etc.
+
+
+### Schema
+
+```yaml
+resources:          # types: pipelines | repositories | containers | packages
```

Comment:
This section needs changes. The resource doesn't need all the imperative stuff around how much to fetch and whether to checkout LFS files. That belongs in `checkout`.

See [docs on what's in the repositories resource](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=vsts&tabs=schema#repository) and [what's in the checkout step](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=vsts&tabs=schema#checkout). I also have a [spec up for multiple checkout](https://github.com/Microsoft/azure-pipelines-yaml/pull/60). If it needs amendments, feel free to comment on the PR.

Reply:
Yeah, now that we are enabling multiple repo checkout support, there are few scenarios we are missing out with your proposal here: https://github.com/Microsoft/azure-pipelines-yaml/blob/3a778fa86fa2536ff64b97bfdb17a13ba4e23d08/design/multicheckout.md

1. User should be aware of the defaults that are taken when a repo is checked out by default. For ex. What is the fetchDepth, are the creds persisted etc. 

2. If user choose the change the default behavior of the checkout, then he/she have to define checkout explicitly across all the jobs for all the repositories. This will be tedious and bloats up the YAML file. 

Moreover, these are optional fields and user doesn't have to provide if not required. And user can always override within the checkout step. By providing support for these details within the resource, the process can get simplified. Please share your thoughts on it.

Reply:
I'm still not convinced. Having more than one "right" way to do things (defaults from the resource vs. overrides on the checkout step) will lead to confusion. I expect templates to address the repetition problem; if we're missing a feature there, let's fix that.

Reply:
These are properties of explicitly how to check out the repo and not about the repo resource itself.  Those properties belong with the checkout step.

Reply:
Okay, makes sense. I will remove the repo properties from resources and will find out if templates can solve this repetition problem.